File: lib/AUTOLQR/FixedPointQ.cpp
#include "FixedPointQ.h"
#include <math.h>
#include <string.h>

namespace fxq {

q_t f2q(float x, int sh, Status* st) {
    double v = (double)x * (double)(1u << sh);
    if (v > 2147483647.0) { st->overflow = true; return INT32_MAX; }
    if (v < -2147483648.0) { st->overflow = true; return INT32_MIN; }
    return (q_t)llround(v);
}

float q2f(q_t x, int sh) {
    return (float)x / (float)(1u << sh);
}

q_t requant(q_t x, int shFrom, int shTo, Status* st) {
    int d = shTo - shFrom;
    if (d == 0) return x;
    if (d > 0) {
        int64_t r = (int64_t)x << d;
        if (r > INT32_MAX) { st->overflow = true; return INT32_MAX; }
        if (r < INT32_MIN) { st->overflow = true; return INT32_MIN; }
        return (q_t)r;
    }
    // d < 0: perde resolução ao arredondar para o shift menor (mesma
    // convenção de truncamento dos demais operadores deste kernel).
    return (q_t)(x >> (-d));
}

// matmul_q/transpose_q/add_q/sub_q agora `inline` em FixedPointQ.h (ver
// comentário lá — recuperar o inlining perdido em 13b33bb).

FXQ_FAST_ATTR bool invert_q(const q_t* src, q_t* dst, int n, int sh, Status* st) {
    q_t aug[12 * 24]; // n até 12 (matrizes 6x6 do laço e o pencil 12x12 do setup do SDA_SS)
    const int n2 = 2 * n;
    const q_t one = f2q(1.0f, sh, st);
    const q_t pivot_floor = f2q(1e-4f, sh, st); // piso > resolução do shift em uso
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) aug[i * n2 + j] = src[i * n + j];
        for (int j = 0; j < n; j++) aug[i * n2 + n + j] = (i == j) ? one : 0;
    }
    for (int i = 0; i < n; i++) {
        q_t maxv = aug[i * n2 + i]; if (maxv < 0) maxv = -maxv;
        int mr = i;
        for (int k = i + 1; k < n; k++) {
            q_t v = aug[k * n2 + i]; if (v < 0) v = -v;
            if (v > maxv) { maxv = v; mr = k; }
        }
        if (maxv < pivot_floor) return false; // ~singular no domínio fixed-point
        if (mr != i)
            for (int j = 0; j < n2; j++) { q_t t = aug[i * n2 + j]; aug[i * n2 + j] = aug[mr * n2 + j]; aug[mr * n2 + j] = t; }
        // Recíproco do pivô calculado UMA vez e multiplicado nas n2 colunas,
        // em vez de n2 divisões — o Xtensa LX7 não tem divisão de 64 bits em
        // hardware (fase 2 da otimização, ver docs/auditoria_solvers_riccati.md
        // Seção 12; MUDA o arredondamento — inv_piv perde a resolução de
        // qdiv(x,piv) direto, então este resultado NÃO é mais bit-a-bit igual
        // ao de antes; revalidado por tolerância contra o scipy, não por
        // igualdade exata).
        q_t piv = aug[i * n2 + i];
        q_t inv_piv = qdiv(one, piv, sh, st);
        for (int j = 0; j < n2; j++) aug[i * n2 + j] = qmul(aug[i * n2 + j], inv_piv, sh, st);
        for (int k = 0; k < n; k++)
            if (k != i) {
                q_t f = aug[k * n2 + i];
                if (f != 0)
                    for (int j = 0; j < n2; j++) aug[k * n2 + j] -= qmul(f, aug[i * n2 + j], sh, st);
            }
    }
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++) dst[i * n + j] = aug[i * n2 + n + j];
    return true;
}

// ----------------------------------------------------------------------------
// Laço de duplicação. n<=6 assumido (buffers de 36 elementos).
// ----------------------------------------------------------------------------
FXQ_FAST_ATTR bool doubling_loop_q(q_t* Ak, q_t* Gk, q_t* Hk, int n, int sh, Variant variant,
                      int maxIterations, int invRelTolerance,
                      float* cum_s_out, Status* st) {
    const int nn = n * n;
    const q_t one = f2q(1.0f, sh, st);
    float cum_s = 1.0f;
    st->iterations = maxIterations;

    q_t Akn[36], Gkn[36], Hkn[36];
    q_t AT[36], V[36], W[36], T1[36], T2[36], T3[36];

    // Somas da ultima iteracao executada, para o passo relativo ser convertido
    // a float UMA vez na parada e nao a cada iteracao (ver bloco de convergencia).
    int64_t ultDiffSq = 0, ultHSq = 0;
    bool ultSaturou = false;

    for (int it = 0; it < maxIterations; it++) {
        if (variant == Variant::AdaptiveScaling) {
            // s = sqrt(||H||/||G||), saturado em [0.1,10], aplicado a G e H.
            // A norma e o fator s em si são calculados em float (poucas
            // chamadas de sqrtf por execução, desprezível frente às 8 matmuls
            // por iteração) — só os dados de Ak/Gk/Hk permanecem inteiros.
            int64_t sumG = 0, sumH = 0;
            for (int i = 0; i < nn; i++) {
                sumG += (int64_t)Gk[i] * (int64_t)Gk[i];
                sumH += (int64_t)Hk[i] * (int64_t)Hk[i];
            }
            float scale = 1.0f / (float)(1LL << sh);
            float normG = sqrtf((float)sumG) * scale;
            float normH = sqrtf((float)sumH) * scale;
            float s_f = 1.0f;
            if (normG > 1e-10f && normH > 1e-10f) {
                s_f = sqrtf(normH / normG);
                s_f = fminf(fmaxf(s_f, 0.1f), 10.0f);
            }
            q_t s_q = f2q(s_f, sh, st);
            for (int i = 0; i < nn; i++) {
                Gk[i] = qmul(Gk[i], s_q, sh, st);
                Hk[i] = qdiv(Hk[i], s_q, sh, st);
            }
            cum_s *= s_f;
        }

        if (variant == Variant::AlternatingVW) {
            // V = (I + Gk·Hk)^-1 ; W = (I + Hk·Gk)^-1 — duas inversões, para
            // medir o efeito da ordem de multiplicação sob quantização (ver
            // docs/auditoria_solvers_riccati.md; em float as duas formas são
            // idênticas por push-through, em Q13.18 não são bit-a-bit iguais).
            matmul_q(Gk, Hk, T1, n, n, n, sh, st);
            for (int i = 0; i < n; i++) T1[i * n + i] += one;
            if (!invert_q(T1, V, n, sh, st)) return false;

            matmul_q(Hk, Gk, T1, n, n, n, sh, st);
            for (int i = 0; i < n; i++) T1[i * n + i] += one;
            if (!invert_q(T1, W, n, sh, st)) return false;

            matmul_q(Ak, V, T1, n, n, n, sh, st);      // T1 = Ak·V
            matmul_q(T1, Ak, Akn, n, n, n, sh, st);    // Akn = Ak·V·Ak

            transpose_q(Ak, AT, n, n);
            matmul_q(Gk, AT, T2, n, n, n, sh, st);     // T2 = Gk·Ak'
            matmul_q(T1, T2, T3, n, n, n, sh, st);     // T3 = Ak·V·Gk·Ak'
            add_q(Gk, T3, Gkn, nn);

            matmul_q(W, Hk, T2, n, n, n, sh, st);      // T2 = W·Hk
            matmul_q(T2, Ak, T3, n, n, n, sh, st);     // T3 = W·Hk·Ak
            matmul_q(AT, T3, T2, n, n, n, sh, st);     // T2 = Ak'·W·Hk·Ak
            add_q(Hk, T2, Hkn, nn);
        } else {
            // Standard (também usado, com o rescale acima, por AdaptiveScaling):
            // V = (I + Gk·Hk)^-1, mesma inversa para as três atualizações.
            matmul_q(Gk, Hk, T1, n, n, n, sh, st);
            for (int i = 0; i < n; i++) T1[i * n + i] += one;
            if (!invert_q(T1, V, n, sh, st)) return false;

            matmul_q(Ak, V, T1, n, n, n, sh, st);      // T1 = Ak·V
            matmul_q(T1, Ak, Akn, n, n, n, sh, st);    // Akn = Ak·V·Ak

            transpose_q(Ak, AT, n, n);
            matmul_q(Gk, AT, T2, n, n, n, sh, st);     // T2 = Gk·Ak'
            matmul_q(T1, T2, T3, n, n, n, sh, st);     // T3 = Ak·V·Gk·Ak'
            add_q(Gk, T3, Gkn, nn);

            matmul_q(V, Ak, T2, n, n, n, sh, st);      // T2 = V·Ak
            matmul_q(Hk, T2, T3, n, n, n, sh, st);     // T3 = Hk·V·Ak
            matmul_q(AT, T3, T2, n, n, n, sh, st);     // T2 = Ak'·Hk·V·Ak
            add_q(Hk, T2, Hkn, nn);
        }

        // Convergência: norma de Frobenius relativa ‖ΔHk‖_F/‖Hk‖_F < tol —
        // MESMA norma do caminho float (AutoLQR.cpp), por pedido explícito
        // do usuário (antes era máx-abs; ver docs/auditoria_solvers_riccati.md,
        // Seção 13/14). O CRITÉRIO não muda: mesma norma, mesmo τ, mesmo
        // orçamento, mesmas decisões. Muda a conta.
        //
        // A Seção 14 daquele documento justificava calcular em float dizendo
        // que a soma inteira "evitaria estourar int64 sem ganho relevante de
        // desempenho, já que o custo é desprezível frente às multiplicações de
        // matriz". O microbenchmark posterior (outputs/serial_norm_benchmark.txt)
        // mediu o contrário: 214,79 us, ou 43,7% da iteração do SDA-fx e 58,7%
        // da iteração de valor, dos quais 129,80 us são só as 72 divisões em
        // software do q2f() (432,7 ciclos cada). A norma inteira custa 5,47 us.
        //
        // Três passos de álgebra tiram a conta do float, sem tocar no critério:
        //   1. elevar ao quadrado os dois lados     -> some a raiz;
        //   2. multiplicar por ‖H‖²                 -> some a divisão;
        //   3. o fator de escala 2^(2s) cancela na razão -> somem as conversões.
        // Sobra  ‖ΔH‖²·τ⁻² < ‖H‖²  em inteiro puro, e EXATO: o caminho antigo
        // arredondava em cada conversão e em cada acumulação (mantissa de 24
        // bits). Onde os dois discordam, quem está certo é este.
#ifdef SDRE_NORMA_FLOAT_LEGADA
        float diffSqF = 0.0f, hSqF = 0.0f;
        bool bitExact = true;
        for (int i = 0; i < nn; i++) {
            if (Hkn[i] != Hk[i]) bitExact = false;
            float d = q2f(Hkn[i], sh) - q2f(Hk[i], sh);
            float h = q2f(Hk[i], sh);
            diffSqF += d * d;
            hSqF += h * h;
        }
        memcpy(Ak, Akn, nn * sizeof(q_t));
        memcpy(Gk, Gkn, nn * sizeof(q_t));
        memcpy(Hk, Hkn, nn * sizeof(q_t));
        float relF = (hSqF > 1e-20f) ? sqrtf(diffSqF / hSqF) : sqrtf(diffSqF);
        st->rel_step = relF;
        st->bit_exact_zero = bitExact;
        if (relF < (1.0f / (float)invRelTolerance)) { st->iterations = it + 1; break; }
#else
        // Teto de acumulação: 36 termos de até (2^31)² somam ~2^67 e estouram o
        // int64. Não acontece na faixa de operação (‖P‖_F ~ 0,43, entradas ~2^17,
        // soma ~2^39), mas o mapa de segurança varre a região de quebra DE
        // PROPÓSITO, e lá acontece. Em vez de supor a faixa, o laço para de somar
        // ao chegar em 2^61 e marca `saturou` — que significa "muito acima de τ",
        // isto é, não convergiu. O `d != 0` continua rodando nos 36 termos para o
        // bit_exact_zero não depender da saturação.
        const int64_t LIM = (int64_t)1 << 61;
        int64_t diffSq = 0, hSq = 0;
        bool bitExact = true, saturou = false;
        for (int i = 0; i < nn; i++) {
            const q_t d = Hkn[i] - Hk[i];
            if (d != 0) bitExact = false;
            if (saturou) continue;
            const int64_t d2 = (int64_t)d * (int64_t)d;
            const int64_t h2 = (int64_t)Hk[i] * (int64_t)Hk[i];
            if (d2 > LIM - diffSq || h2 > LIM - hSq) { saturou = true; continue; }
            diffSq += d2;
            hSq += h2;
        }
        memcpy(Ak, Akn, nn * sizeof(q_t));
        memcpy(Gk, Gkn, nn * sizeof(q_t));
        memcpy(Hk, Hkn, nn * sizeof(q_t));

        ultDiffSq = diffSq;
        ultHSq = hSq;
        ultSaturou = saturou;
        st->bit_exact_zero = bitExact;

        bool convergiu = false;
        if (!saturou) {
            // τ⁻² = invRelTolerance². Se o produto estoura, o passo está ordens
            // de grandeza acima de τ: nenhum hSq representável o alcançaria,
            // logo não convergiu — e não se forma o produto.
            const int64_t invTol2 = (int64_t)invRelTolerance * (int64_t)invRelTolerance;
            int64_t lhs;
            convergiu = !__builtin_mul_overflow(diffSq, invTol2, &lhs) && (lhs < hSq);
        }
        if (convergiu) { st->iterations = it + 1; break; }
#endif
    }

#ifndef SDRE_NORMA_FLOAT_LEGADA
    // O passo relativo é instrumentação (Fig. 3(c) do artigo), lido só na parada.
    // Convertido aqui UMA vez, em vez das 72 conversões por iteração que o
    // caminho antigo pagava nove vezes por solve.
    st->rel_step = (!ultSaturou && ultHSq > 0)
                       ? sqrtf((float)ultDiffSq / (float)ultHSq)
                       : 1.0f;
#endif

    if (cum_s_out) *cum_s_out = cum_s;
    return true;
}

} // namespace fxq

File: python/fixedpoint_q.py
"""
Emulação em Python (inteiros nativos, precisão arbitrária, com os mesmos
limites explícitos de int32/int64) do kernel Q13.18 de
lib/AUTOLQR/FixedPointQ.{h,cpp} — usada para prever e explicar o erro de
quantização dos seis solvers `_FIXED` fora do alvo, e para gerar a figura
central do artigo (erro de K vs. faixa dinâmica ocupada, com o teto ±8192
marcado). Não substitui a verificação no hardware real: aqui apenas se
reproduz a ARITMÉTICA (mesmas operações, mesmos arredondamentos, mesmos
critérios de overflow); os limites de RAM/stack do ESP32-S2 não têm
equivalente aqui.

Cada função espelha, linha por linha, a correspondente em FixedPointQ.cpp —
ver esse arquivo para o porquê de cada escolha (pivot_floor, saturação em
[0,1;10] no ASDA, a forma V/W separada do ADDA etc.).
"""

import numpy as np

Q_SHIFT_DEFAULT = 18
INT32_MAX = 2**31 - 1
INT32_MIN = -2**31


class Status:
    def __init__(self):
        self.overflow = False
        self.iterations = 0
        self.max_abs_seen = 0
        self.rel_step = 1.0        # espelha Status::rel_step do C++
        self.bit_exact_zero = False


def f2q(x, sh, st):
    v = float(x) * float(1 << sh)
    if v > 2147483647.0:
        st.overflow = True
        return INT32_MAX
    if v < -2147483648.0:
        st.overflow = True
        return INT32_MIN
    return int(round(v))


def q2f(x, sh):
    return float(x) / float(1 << sh)


def qmul(a, b, sh, st):
    r = (a * b) >> sh
    if r > INT32_MAX:
        st.overflow = True
        return INT32_MAX
    if r < INT32_MIN:
        st.overflow = True
        return INT32_MIN
    return r


def qdiv(a, b, sh, st):
    if b == 0:
        st.overflow = True
        return INT32_MAX if a >= 0 else INT32_MIN
    # divisao truncada em direcao a zero, igual ao C++ (int64_t / int64_t) —
    # o // do Python trunca para -infinito, por isso o _trunc_div dedicado.
    num = a << sh
    r = _trunc_div(num, b)
    if r > INT32_MAX:
        st.overflow = True
        return INT32_MAX
    if r < INT32_MIN:
        st.overflow = True
        return INT32_MIN
    return r


def _trunc_div(num, den):
    q = abs(num) // abs(den)
    if (num < 0) != (den < 0):
        q = -q
    return q


def matmul_q(a, b, r1, c1, c2, sh, st):
    """a: lista/array (r1*c1), b: (c1*c2) -> retorna lista (r1*c2)."""
    c = [0] * (r1 * c2)
    for i in range(r1):
        for j in range(c2):
            acc = 0
            for k in range(c1):
                acc += a[i * c1 + k] * b[k * c2 + j]
            r = acc >> sh
            if r > INT32_MAX:
                st.overflow = True
                r = INT32_MAX
            elif r < INT32_MIN:
                st.overflow = True
                r = INT32_MIN
            c[i * c2 + j] = r
            av = abs(r)
            if av > st.max_abs_seen:
                st.max_abs_seen = av
    return c


def transpose_q(a, r, c):
    at = [0] * (r * c)
    for i in range(r):
        for j in range(c):
            at[j * r + i] = a[i * c + j]
    return at


def add_q(a, b):
    return [x + y for x, y in zip(a, b)]


def sub_q(a, b):
    return [x - y for x, y in zip(a, b)]


def invert_q(src, n, sh, st):
    """Gauss-Jordan com pivotamento parcial em Q-format — espelho de
    invert_q() em FixedPointQ.cpp. Retorna (dst, ok)."""
    n2 = 2 * n
    one = f2q(1.0, sh, st)
    pivot_floor = f2q(1e-4, sh, st)
    aug = [0] * (n * n2)
    for i in range(n):
        for j in range(n):
            aug[i * n2 + j] = src[i * n + j]
        for j in range(n):
            aug[i * n2 + n + j] = one if i == j else 0

    for i in range(n):
        maxv = abs(aug[i * n2 + i])
        mr = i
        for k in range(i + 1, n):
            v = abs(aug[k * n2 + i])
            if v > maxv:
                maxv, mr = v, k
        if maxv < pivot_floor:
            return None, False
        if mr != i:
            for j in range(n2):
                aug[i * n2 + j], aug[mr * n2 + j] = aug[mr * n2 + j], aug[i * n2 + j]
        # Reciproco do pivo calculado UMA vez e multiplicado nas n2 colunas --
        # espelha FixedPointQ.cpp:61-63 (o Xtensa LX7 nao tem divisao de 64
        # bits em hardware, entao o C++ troca n2 divisoes por 1 divisao + n2
        # multiplicacoes). MUDA o arredondamento em relacao a qdiv direto, por
        # isso tem de ser replicado aqui para o emulador continuar fiel.
        piv = aug[i * n2 + i]
        inv_piv = qdiv(one, piv, sh, st)
        for j in range(n2):
            aug[i * n2 + j] = qmul(aug[i * n2 + j], inv_piv, sh, st)
        for k in range(n):
            if k != i:
                f = aug[k * n2 + i]
                if f != 0:
                    for j in range(n2):
                        aug[k * n2 + j] -= qmul(f, aug[i * n2 + j], sh, st)

    dst = [0] * (n * n)
    for i in range(n):
        for j in range(n):
            dst[i * n + j] = aug[i * n2 + n + j]
    return dst, True


VARIANT_STANDARD = "standard"
VARIANT_ADAPTIVE_SCALING = "adaptive_scaling"
VARIANT_ALTERNATING_VW = "alternating_vw"


def doubling_loop_q(Ak, Gk, Hk, n, sh, variant, max_iterations, inv_rel_tolerance, st):
    """Espelho de doubling_loop_q() em FixedPointQ.cpp. Ak,Gk,Hk: listas
    planas n*n (inteiros Q-format), modificadas e retornadas (não in-place,
    ao contrário do C++, para evitar armadilhas de aliasing em Python).
    Retorna (Ak, Gk, Hk, cum_s, ok)."""
    nn = n * n
    one = f2q(1.0, sh, st)
    cum_s = 1.0
    st.iterations = max_iterations

    Ak = list(Ak)
    Gk = list(Gk)
    Hk = list(Hk)

    # somas da ultima iteracao executada, para o passo relativo sair uma vez
    ult_diff_sq, ult_h_sq, ult_saturou = 0, 0, False

    for it in range(max_iterations):
        if variant == VARIANT_ADAPTIVE_SCALING:
            sumG = sum(g * g for g in Gk)
            sumH = sum(h * h for h in Hk)
            scale = 1.0 / float(1 << sh)
            normG = (sumG ** 0.5) * scale
            normH = (sumH ** 0.5) * scale
            s_f = 1.0
            if normG > 1e-10 and normH > 1e-10:
                s_f = (normH / normG) ** 0.5
                s_f = min(max(s_f, 0.1), 10.0)
            s_q = f2q(s_f, sh, st)
            Gk = [qmul(g, s_q, sh, st) for g in Gk]
            Hk = [qdiv(h, s_q, sh, st) for h in Hk]
            cum_s *= s_f

        if variant == VARIANT_ALTERNATING_VW:
            T1 = matmul_q(Gk, Hk, n, n, n, sh, st)
            for i in range(n):
                T1[i * n + i] += one
            V, ok = invert_q(T1, n, sh, st)
            if not ok:
                return Ak, Gk, Hk, cum_s, False

            T1b = matmul_q(Hk, Gk, n, n, n, sh, st)
            for i in range(n):
                T1b[i * n + i] += one
            W, ok = invert_q(T1b, n, sh, st)
            if not ok:
                return Ak, Gk, Hk, cum_s, False

            AV = matmul_q(Ak, V, n, n, n, sh, st)
            Akn = matmul_q(AV, Ak, n, n, n, sh, st)

            AT = transpose_q(Ak, n, n)
            GAT = matmul_q(Gk, AT, n, n, n, sh, st)
            AVGAT = matmul_q(AV, GAT, n, n, n, sh, st)
            Gkn = add_q(Gk, AVGAT)

            WH = matmul_q(W, Hk, n, n, n, sh, st)
            WHA = matmul_q(WH, Ak, n, n, n, sh, st)
            ATWHA = matmul_q(AT, WHA, n, n, n, sh, st)
            Hkn = add_q(Hk, ATWHA)
        else:
            T1 = matmul_q(Gk, Hk, n, n, n, sh, st)
            for i in range(n):
                T1[i * n + i] += one
            V, ok = invert_q(T1, n, sh, st)
            if not ok:
                return Ak, Gk, Hk, cum_s, False

            AV = matmul_q(Ak, V, n, n, n, sh, st)
            Akn = matmul_q(AV, Ak, n, n, n, sh, st)

            AT = transpose_q(Ak, n, n)
            GAT = matmul_q(Gk, AT, n, n, n, sh, st)
            AVGAT = matmul_q(AV, GAT, n, n, n, sh, st)
            Gkn = add_q(Gk, AVGAT)

            VA = matmul_q(V, Ak, n, n, n, sh, st)
            HVA = matmul_q(Hk, VA, n, n, n, sh, st)
            ATHVA = matmul_q(AT, HVA, n, n, n, sh, st)
            Hkn = add_q(Hk, ATHVA)

        # Frobenius relativa, igual ao C++ (FixedPointQ.cpp:doubling_loop_q).
        # O CRITERIO nao mudou; a conta sim: ‖ΔH‖²·τ⁻² < ‖H‖² em inteiro, sem
        # conversao, raiz nem divisao. O fator de escala 2^(2s) cancela na razao.
        # ESTE ESPELHO TEM DE SEGUIR O C++ EXATAMENTE: malha_fechada_trajetorias.py
        # e bench_trajetorias.py simulam no host o que a placa executa, e qualquer
        # divergencia aqui vira divergencia entre simulacao e firmware.
        #
        # Python tem inteiro de precisao arbitraria, entao nao ha' o teto de
        # 2^61 do int64 do C++ -- mas o `saturou` e' replicado assim mesmo, para
        # que as duas implementacoes tomem a MESMA decisao na regiao de quebra
        # que o mapa de seguranca varre.
        LIM = 1 << 61
        diff_sq = 0
        h_sq = 0
        bit_exact = True
        saturou = False
        for i in range(nn):
            d = Hkn[i] - Hk[i]
            if d != 0:
                bit_exact = False
            if saturou:
                continue
            d2 = d * d
            h2 = Hk[i] * Hk[i]
            if d2 > LIM - diff_sq or h2 > LIM - h_sq:
                saturou = True
                continue
            diff_sq += d2
            h_sq += h2

        Ak, Gk, Hk = Akn, Gkn, Hkn
        st.bit_exact_zero = bit_exact
        ult_diff_sq, ult_h_sq, ult_saturou = diff_sq, h_sq, saturou

        convergiu = False
        if not saturou:
            inv_tol2 = inv_rel_tolerance * inv_rel_tolerance
            lhs = diff_sq * inv_tol2
            # o C++ recusa por overflow do int64 o que aqui nunca estoura; a
            # condicao equivalente e' o produto passar do que um int64 guarda.
            convergiu = lhs <= 0x7FFFFFFFFFFFFFFF and lhs < h_sq
        if convergiu:
            st.iterations = it + 1
            break

    # passo relativo: instrumentacao, convertido uma vez na parada (igual ao C++)
    st.rel_step = ((ult_diff_sq / ult_h_sq) ** 0.5
                   if (not ult_saturou and ult_h_sq > 0) else 1.0)

    return Ak, Gk, Hk, cum_s, True


# ---------------------------------------------------------------------------
# Conveniência: matriz numpy <-> lista plana Q-format
# ---------------------------------------------------------------------------
def mat_to_q(M, sh, st):
    return [f2q(v, sh, st) for v in np.asarray(M, dtype=float).flatten()]


def q_to_mat(vals, n, m, sh):
    return np.array([q2f(v, sh) for v in vals], dtype=float).reshape(n, m)


if __name__ == "__main__":
    # Auto-teste rápido: qmul/qdiv/invert_q em uma matriz 3x3 bem condicionada,
    # conferindo contra numpy.linalg.inv dentro da resolução do Q13.18.
    st = Status()
    sh = Q_SHIFT_DEFAULT
    M = np.array([[4.0, 1.0, 0.0], [1.0, 3.0, 1.0], [0.0, 1.0, 2.0]])
    Mq = mat_to_q(M, sh, st)
    dst, ok = invert_q(Mq, 3, sh, st)
    Minv_q = q_to_mat(dst, 3, 3, sh)
    Minv_ref = np.linalg.inv(M)
    erro = np.max(np.abs(Minv_q - Minv_ref))
    print("invert_q ok=%s overflow=%s erro_max=%.3e (resolucao Q13.18=%.3e)" % (
        ok, st.overflow, erro, 2.0 ** -18))
    assert ok and not st.overflow and erro < 1e-3
    print("auto-teste OK")
